build-system-abstraction.rst

```diff
@@ -0,0 +1,419 @@
+    description is presumed to be identical.
+
+Python environments and hermetic builds
+---------------------------------------
+
+This specification does not prescribe whether builds should be hermetic or not.
+Existing build tools like setuptools will use installed versions of build time
+requirements (e.g. setuptools_scm) and only install other versions on version
+conflicts or missing dependencies. However its likely that better consistency
+can be created by always isolation builds and using only the specified dependencies.
+
+However there are nuanced problems there - such as how can users force the
+avoidance of a bad version of a build requirement which meets some packages
+dependencies. Future PEPs may tackle this problem, but it is not currently in
+scope - it does not affect the metadata required to coordinate between build
+systems and things that need to do builds.
+
+Upgrades
+--------
+
+Both 'pypa.json' and the build tool description are versioned to permit future
+incompatible changes. There is a sequence dependency here.
+
+Upgrades to the schemas defined in this specification must proceed with the
+consumers first. To ensure that consumers should refuse to operate when
+'pypa.json' has a schema version that they do not recognise.
+
+Build tools listed in a 'pypa.json' with schema version 1 must not generate a
+build tool description with a version other than 1 (or absent, as the default
+is 1).
+
+Thus the sequence for upgrading either of schemas in a new PEP will be:
+
+1. Issue new PEP defining build tool description and 'pypa.json' schemas. The
+   'pypa.yaml' schema version must change if either the 'pypa.yaml' schema has
+   changed or the build tool description schema has changed. The build tool
+   description schema version must change if the build tool description schema
+   has changed.
+2. Consumers (e.g. pip) implement support for the new schema version.
+3. Build tool authors implement the new schemas, and publish updated reference
+   'pypa.json' files for their users.  4. Package authors opt into the new
+   schema when they are happy to introduce a dependency on the version of
+   'pip' (and potentially other consumers) that introduced support for the new
+   schema version.
+
+The *same* process will take place for the initial deployment of this PEP:-
+the propogation of the capability to use this PEP without a `setuptools shim`_
+will be largely gated by the adoption rate of the first version of pip that
+supports it.
+
+Static metadata in sdists
+-------------------------
+
+This PEP does not tackle the current inability to trust static metadata in
+sdists. That is a separate problem to identifying and consuming the build
+system that is in use in a source tree, whether it came from an sdist or not.
+
+Handling of compiler options
+----------------------------
+
+Handling of different compiler options is out of scope for this specification.
```

Comment:
Further, the whole point of them being out of scope is that we're not defining their behaviour: we're saying that tools like conda or pip that choose to invoke build systems can build their own UI and logic around this, we're merely calling out that this is likely going to happen, and that what is passed where needs to understand different build systems will have different things here.